### PR TITLE
Infer `primary_key: :id` on associations with composite primary key models

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
### Motivation / Background

Previously, a composite primary key model would need to specify either a `primary_key` or `query_constraints` option on its associations. Without it, a `CompositePrimaryKeyMismatchError` would be raised.

Most of the time, the composite primary key includes the `:id` column, and associations already expect that `:id` is being used as the primary key for the association (and the associated model is inferring a foreign key of `<cpk_model>_id`).

Rather than requiring users to define `primary_key: :id` throughout an application with composite primary key associations, we can infer that the `:id` column should be used. Users can still override this behaviour by specifying `primary_key:` or `query_constraints:` on the association.

Note that, if the composite primary key for a model does _not_ include `:id`, Rails **won't** infer the primary key for any related associations, and users must still specify `query_constraints` or `primary_key`.

Prior to this change, you'd need to do the following to set up associations for composite primary key models:

```ruby
class Order
  self.primary_key = [:shop_id, :id]

  has_many :order_agreements, primary_key: :id
end

class OrderAgreement
  belongs_to :order, primary_key: :id
end
```

After this change, the `primary_key` option no longer needs to be specified:

```ruby
class Order
  self.primary_key = [:shop_id, :id]

  has_many :order_agreements
end

class OrderAgreement
  belongs_to :order
end
```
### Detail

Most of the changes in this PR involve inferring the `primary_key` for an association, when a) the owner in the association is a composite primary key model, and b) the composite primary key includes `:id`.

I ended up refactoring `#save_has_one_association` to use `#compute_primary_key` so it could share this PK-inference logic with `#save_belongs_to_association`.

I also needed to change some of the models that were testing logic related to `CompositePrimaryKeyMismatchError`, since these associations are no longer malformed (as Active Record can now infer the primary_key on those associations properly). I opted to change the primary key on those "broken" models to exclude the `id` column, since that prevents Rails from inferring the PK on the association, and causes the expected `CompositePrimaryKeyMismatchError` to be raised.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Tests are added or updated if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/lib/active_record/reflection.rb [hunk 1]
```diff
@@ -533,6 +532,10 @@ def active_record_primary_key
           end
         elsif active_record.has_query_constraints? || options[:query_constraints]
           active_record.query_constraints_list
+        elsif active_record.composite_primary_key?
+          # If active_record has composite primary key of shape [:<tenant_key>, :id], infer primary_key as :id
+          primary_key = primary_key(active_record)
+          primary_key.include?("id") ? "id".freeze : primary_key.freeze
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/reflection.rb:538]: Don't need to freeze here.

**Comment by eileencodes** [comment]
@adrianna-chang-shopify can you rebase this with `main`? globalid was locked so that will fix the test failures here.

**Review by nvasilevski — APPROVED** [review]
👍

**Comment by adrianna-chang-shopify** [comment]
👋 Hey @medranocalvo , thanks for your feedback!

> The inference implemented by this change codifies the following assumption:

Our perspective is that this is a fair assumption to make. The idea with this change was to make it easy for models with a primary key of `id` to move to a composite primary key (ie. in an application that begins moving its models to use tenant-based sharding) while maintaining their existing associations without having to add `query_constraints` or specify the `primary_key` option on their associations.

In general, Rails assumes that an id column on a table should be a unique identifier, so what you've said here holds true:
> That is, is the :id attribute always assumed to be an unique key, irrespective of the value of self.primary_key?

Rails' recommendation for models with composite primary keys is to avoid using `:id` for any of the columns in the primary key when possible. New tables should pick a different column name, like `oid`.

> I think that Rails should infer that the primary key for the association is the target's primary key; in the example above primary_key [:shop_id, :id].

I agree that if it were possible to infer the composite foreign key for all associations based on the owner's primary key, this would be ideal (and better than what we've done in this PR), but the problem is that inferring the full composite key is really tricky for Rails to do. [@eileencodes  talks about it a bit in this PR,](https://github.com/rails/rails/pull/49088) but the TL;DR is that it's challenging to know which parts of the primary key resolve to which foreign key columns. We also might not want to assume that for every CPK model, all parts of its composite primary key need to be specified as foreign key columns on every associated model. **So to be clear**: the current approach is not an alternative for a "smarter" resolution strategy that would take into account the full CPK (which would be better, but is hard to do!). Instead, it's an alternative to "not inferring anything" and forcing existing associations that have a PK-FK setup with `id` to change all their associations.

The TL;DR is that we assume the following:
* If multiple columns need to be used as the foreign key for an association, the user must specify query constraints, because it's too hard for Rails to derive this automatically
* Rails can handle the simple use case of deriving an `:id` primary key for CPK associations where `:id` is a part of the primary key, with the assumption that `:id` is the column that specifies the PK-FK relationship. If this is not the case, users must specify the `query_constraints:` option.
